torch/_subclasses/meta_utils.py

```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
                         r = base.as_strided(sizes, strides, sym(t.storage_offset()))
                 else:
                     is_leaf = safe_is_leaf(t)
+                    sizes, strides = sym_sizes_strides(t)
+                    storage_offset = sym(t.storage_offset())
                     # Fake up some autograd history.
                     if t.requires_grad:
-                        r = torch.empty(
-                            (0,), dtype=t.dtype, device="meta", requires_grad=True
+                        r = callback(
+                            lambda: torch.empty_strided(
+                                sizes, strides, dtype=t.dtype, device="meta"
+                            )
                         )
+                        assert safe_is_leaf(
+                            r
+                        ), "the callback you passed in doesn't detach"
+                        r.requires_grad = t.requires_grad
```

Comment:
Now that this is moved outside of the empty call, the logic here is:
```
if ..:
  do A
  do B
else:
  do A
```
We can clean that up.